### PR TITLE
First Party Data module: Add new module and two submodules to populate defaults and validate ortb2

### PR DESCRIPTION
**Pull request by mmoschovas**
## Type of change

- [ ] Bugfix
- [X ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

Create new optional module to validate ortb2 as well as set ortb2 defaults if applicable.
- verify OpenRTB datatypes, remove/warn any that are likely to choke downstream readers
- verify that certain OpenRTB attributes are not specified
- suppress user FPD based on a TBD opt-out signal (_pubcid_optout)
- populate available data into object: referer, meta-keywords, cur, etc.

Issue described: https://github.com/prebid/Prebid.js/issues/6099

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

File: test/spec/modules/firstPartyData_spec.js
Comment (Fawke) [inline test/spec/modules/firstPartyData_spec.js:5]: We are not using the imports, better to remove them.

File: test/spec/modules/firstPartyData_spec.js
Comment (Fawke) [inline test/spec/modules/firstPartyData_spec.js:17]: `logErrorSpy` variable is not used anywhere in this file...better remove it?

File: modules/firstPartyData/index.js
Comment (Fawke) [inline modules/firstPartyData/index.js:254]: Should there be a check in place to see if publisher has defined `pbjs.setConfig({ ortb2: { ... } })` before we start execution of the init() function?

In some cases, the module may've been included but, the publisher has not defined **ortb2** property, then it doesn't make sense to run these checks.

File: modules/firstPartyData/index.js
Comment (Fawke) [inline modules/firstPartyData/index.js:63]: The spec [here](https://github.com/prebid/Prebid.js/issues/6099#issue-759814143) specifies `ortb2.device.h` and `ortb2.device.w` but here, `ortb2.device.height` and `ortb2.device.width` is used.

**Review by Fawke — COMMENTED** [review]
Hi @mmoschovas,

Thanks for creating this PR, the code mostly looks fine to me apart from the few doubts I had which I've mentioned as inline review comments.

I haven't tested the module end to end yet, I was wondering if you'd be able to provide a test page which you may've used for your internal testing.

File: modules/firstPartyData/index.js
Comment (mmoschovas) [reply modules/firstPartyData/index.js:254]: Part of the functionality is validation; however, the other piece here is that default values are set i.e. device data, referer, etc. Meaning that the existence of ortb2 is not required for the module to run.

File: modules/firstPartyData/index.js
Comment (mmoschovas) [reply modules/firstPartyData/index.js:63]: Fixing this. Thanks for catching!

File: test/spec/modules/firstPartyData_spec.js
Comment (mmoschovas) [reply test/spec/modules/firstPartyData_spec.js:17]: Removing

File: modules/firstPartyData/index.js
Comment (Fawke) [reply modules/firstPartyData/index.js:267]: I agree with what you're saying. Earlier I misunderstood this feature to perform checks on ortb2 only if it's present. But regardless of this presence, if this module is present, it'll anyways create a ortb2 object with default values.

**Comment by Fawke** [comment]
Hi @mmoschovas,

I was trying to test this module according to the spec documented [here](https://github.com/prebid/Prebid.js/issues/6099#issue-759814143). I needed a little clarification on points 11 and 13.

**11** says
>The module should provide a function to force a refresh of enrichments and validations in case the publisher's page can change them after the first auction.

I didn't find any such function.

and **13**, which says
>The following OpenRTB field validations should be performed on global and per-bidder ortb2. If any of them fails, remove the field from the ortb2 data and log a console error.

>site.{name,domain,page,ref,keywords,search} must be strings
>site.{cat, sectioncat, pagecat} must be arrays of strings
>site.{content,publisher} must be an objects
>user.yob must be an integer
>user.{gender, keywords} must be a string

I didn't see this working. I used a non string value for `site.name` and `site.domain` and a string value for `user.yob`, it didn't log any warning, nor did it remove those faulty values from the config. Am I missing something, or doing something wrong while testing it, or you haven't yet implement these features. 

Also, it'd be nice to get rid of the LGTM error if you can. Thanks.

**Comment by mmoschovas** [comment]
@Fawke some of these requirements were added after the PR was originally made. I will be pushing some updates today hopefully to cover these instances.

File: modules/firstPartyData/index.js
Comment (mmoschovas) [reply modules/firstPartyData/index.js:21]: Correct. So left to right in terms of priority. This line is actually just setting the modules global variable which is later merged with the response of getConfig('ortb2) || {}. So if the value exists in ortb2 already it will overwrite the modules 'enrichments' setting and use what was already there.

**Comment by bretg** [comment]
@mmoschovas - based on Patrick's feedback, please update to not set cur or support optout for now. I've added flags to requirements 6 and 9 above to take them out of the initial release.

**Comment by Fawke** [comment]
> @Fawke some of these requirements were added after the PR was originally made. I will be pushing some updates today hopefully to cover these instances.

@mmoschovas Hey, I wanted to review this again, but before that just wanted to confirm if this PR is complete or still WIP. Thanks.

**Comment by mmoschovas** [comment]
@Fawke this should be good to review. No recent changes have been proposed here. Thanks

**Review by Fawke — APPROVED** [review]
LGTM

File: modules/firstPartyData/config.js
Comment (bretg) [inline modules/firstPartyData/config.js:93]: please set optoutApplies for yob and gender

File: modules/firstPartyData/index.js
Comment (bretg) [inline modules/firstPartyData/index.js:313]: Not sure this is right -- what if the publisher adds their own ortb2 data from the page? This module needs to preserve it and merge in the enrichments.

File: modules/firstPartyData/index.js
Comment (bretg) [inline modules/firstPartyData/index.js:307]: I don't want the enrichments done every auction. They should be done only on the initial run (if configured) and on refreshFPD (if configured). Will update the issue.

**Review by bretg — CHANGES_REQUESTED** [review]

File: modules/firstPartyData/index.js
Comment (mmoschovas) [reply modules/firstPartyData/index.js:307]: As discussed the hook is removed after its initial run. The module will only rerun on a subsequent auction if the refreshFPD is triggered

File: modules/firstPartyData/index.js
Comment (mmoschovas) [reply modules/firstPartyData/index.js:313]: This reset is for the ortb2 object within the module itself. Not the global one used in setConfig(...). This is to allow enrichments to be properly reevaluated on a subsequent auction if refreshFPD is triggered prior

File: modules/firstPartyData/config.js
Comment (mmoschovas) [reply modules/firstPartyData/config.js]: Done

**Comment by bretg** [comment]
Sorry for the requirements change here, but after realizing that the validation code takes about 1KB, but would rarely be useful beyond the occasional site change,  the taxonomy committee would like to split the functionality:

1) remove the validation and getConfig({firstPartyData}) code from this PR
2) the external name of this module becomes First Party Data Enrichment
3) move the validation code to a separate PR for a First Party Data Validation module

**Comment by bretg** [comment]
Docs PR https://github.com/prebid/prebid.github.io/pull/2910 -- this assumes that the module is enrichment-only.

**Comment by Fawke** [comment]
Reviewing this again since this module was split into two submodules. Everything seems to be working fine after this change and my original `LGTM` still holds. @bretg Do we another doc for the `validation module` because this PR includes both the submodules.

File: modules/fpdModule/index.md
Comment (bretg) [inline modules/fpdModule/index.md:28]: Let's call this section "Module Control" so as not to confuse it with the user pubcid opt-out. Thanks.

**Review by bretg — CHANGES_REQUESTED** [review]

**Review by bretg — APPROVED** [review]

**Pull request merged**
